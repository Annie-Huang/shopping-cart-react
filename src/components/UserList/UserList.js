import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import toastr from 'toastr';
import * as userActions from '../../actions/userActions';
import * as shoppingCartActions from '../../actions/shoppingCartActions';
import PriceRuleList from './components/PriceRuleList/PriceRuleList';
import Collapse from '../Common/Collapse/Collapse';

export class UserList extends Component {
    selectUser = (event) => {
        this.props.loadUser(event.target.value);
    };

    reset = () => {
        this.props.resetUser();
        this.props.emptyItemInCart();
        toastr.success(`Reset user success`);
    };

    render() {
        const userIds = ['Apple', 'Ford', 'Nike', 'Unilever', 'Default'].map(userId =>
            <button key={userId}
                    value={userId}
                    type="button"
                    className="btn btn-outline-primary"
                    disabled={this.props.buttonDisable}
                    onClick={this.selectUser}
            >
                {userId}
            </button>
        );

        const { id, name, pricingRules } = this.props.user;
        const { products } = this.props;

        return (
            <div>
                <Collapse expanded={!id}>
                    <h2>Please select a user</h2>
                </Collapse>
                <div className='row'>
                    <div className='col-md-6'>
                        <div className="d-flex flex-row justify-content-between my-flex-container">
                            {userIds}
                            <button type="button" className="btn btn-outline-danger" onClick={this.reset}>Reset</button>
                        </div>
                    </div>
                </div>

                <br/>
                {id &&
                    <div>
                        <div className='row'>
                            <div className='col-md-12'>
                                <h2>
                                    Welcome {name}, please pick your Ad from our wonderful selection...
                                </h2>
                            </div>
                        </div>
                        <PriceRuleList products={products} rules={pricingRules}/>
                    </div>
                }
                <br/>
            </div>
        )
    }
}

UserList.propTypes = {
    products: PropTypes.array.isRequired,
    user: PropTypes.object.isRequired,
    buttonDisable: PropTypes.bool.isRequired,
    loadUser: PropTypes.func.isRequired,
    resetUser: PropTypes.func.isRequired,
    emptyItemInCart: PropTypes.func.isRequired,
};

const mapStateToProps = (state) => ({
    products: state.products,
    user: state.user,
    buttonDisable: !!state.user.id,
});

const mapDispatchToProps = (dispatch) => ({
    loadUser: userId => dispatch(userActions.loadUser(userId)),
    resetUser: () => dispatch(userActions.resetUser()),
    emptyItemInCart: () => dispatch(shoppingCartActions.emptyItemInCart()),
});

export default connect(mapStateToProps, mapDispatchToProps)(UserList);
