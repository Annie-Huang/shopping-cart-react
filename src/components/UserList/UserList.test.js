import {shallow} from 'enzyme/build';
import {UserList} from "./UserList";
import React from "react";

describe('UserList presentation', () => {
    const loadUser = jest.fn();
    const resetUser = jest.fn();
    const emptyItemInCart = jest.fn();
    const products = require('../../resources/fixtures/products');
    const user = {};
    const buttonDisable = false;
    const props = {
        loadUser,
        resetUser,
        emptyItemInCart,
        products,
        user,
        buttonDisable
    };

    it('should render UserList component when no user selected', () => {
        expect(shallow(<UserList {...props} />)).toMatchSnapshot();
    });

    it('should render UserList component when a user selected', () => {
        const appleUser =  require('../../resources/fixtures/user-apple');
        const newProps = {
            loadUser,
            resetUser,
            emptyItemInCart,
            products,
            user: appleUser,
            buttonDisable: true
        };
        expect(shallow(<UserList {...newProps} />)).toMatchSnapshot();
    });

    it('#selectUser -- called directly', () => {
        const wrapper = shallow(<UserList {...props} />).instance();
        const event = {
            target: {
                value: 'Apple'
            }
        };
        wrapper.selectUser(event);

        expect(loadUser).toBeCalledWith('Apple');
    });

    it('#reset -- called directly', () => {
        const wrapper = shallow(<UserList {...props} />).instance();
        wrapper.reset();

        expect(resetUser).toHaveBeenCalled();
        expect(emptyItemInCart).toHaveBeenCalled();
    });

    it('#reset should be called when clicking reset button', () => {
        const wrapper = shallow(<UserList {...props} />);

        const buttons = wrapper.find('button');
        const resetButton =  buttons.at(5);
        expect(resetButton.text()).toBe('Reset');
        resetButton.simulate('click');

        expect(resetUser).toHaveBeenCalled();
        expect(emptyItemInCart).toHaveBeenCalled();
    });
});
