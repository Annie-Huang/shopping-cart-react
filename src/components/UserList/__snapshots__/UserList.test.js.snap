// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserList presentation should render UserList component when a user selected 1`] = `
<div>
  <Collapse
    expanded={false}
  >
    <h2>
      Please select a user
    </h2>
  </Collapse>
  <div
    className="row"
  >
    <div
      className="col-md-6"
    >
      <div
        className="d-flex flex-row justify-content-between my-flex-container"
      >
        <button
          className="btn btn-outline-primary"
          disabled={true}
          key="Apple"
          onClick={[Function]}
          type="button"
          value="Apple"
        >
          Apple
        </button>
        <button
          className="btn btn-outline-primary"
          disabled={true}
          key="Ford"
          onClick={[Function]}
          type="button"
          value="Ford"
        >
          Ford
        </button>
        <button
          className="btn btn-outline-primary"
          disabled={true}
          key="Nike"
          onClick={[Function]}
          type="button"
          value="Nike"
        >
          Nike
        </button>
        <button
          className="btn btn-outline-primary"
          disabled={true}
          key="Unilever"
          onClick={[Function]}
          type="button"
          value="Unilever"
        >
          Unilever
        </button>
        <button
          className="btn btn-outline-primary"
          disabled={true}
          key="Default"
          onClick={[Function]}
          type="button"
          value="Default"
        >
          Default
        </button>
        <button
          className="btn btn-outline-danger"
          onClick={[Function]}
          type="button"
        >
          Reset
        </button>
      </div>
    </div>
  </div>
  <br />
  <div>
    <div
      className="row"
    >
      <div
        className="col-md-12"
      >
        <h2>
          Welcome 
          Apple
          , please pick your Ad from our wonderful selection...
        </h2>
      </div>
    </div>
    <PriceRuleList
      products={
        Array [
          Object {
            "id": "classic",
            "name": "Classic Ad",
            "price": 269.99,
          },
          Object {
            "id": "standout",
            "name": "Standout Ad",
            "price": 322.99,
          },
          Object {
            "id": "premium",
            "name": "Premium Ad",
            "price": 394.99,
          },
        ]
      }
      rules={
        Array [
          Object {
            "attributes": Array [
              Object {
                "name": "newPrice",
                "value": 299.99,
              },
            ],
            "productId": "standout",
            "ruleName": "newUnitPrice",
          },
        ]
      }
    />
  </div>
  <br />
</div>
`;

exports[`UserList presentation should render UserList component when no user selected 1`] = `
<div>
  <Collapse
    expanded={true}
  >
    <h2>
      Please select a user
    </h2>
  </Collapse>
  <div
    className="row"
  >
    <div
      className="col-md-6"
    >
      <div
        className="d-flex flex-row justify-content-between my-flex-container"
      >
        <button
          className="btn btn-outline-primary"
          disabled={false}
          key="Apple"
          onClick={[Function]}
          type="button"
          value="Apple"
        >
          Apple
        </button>
        <button
          className="btn btn-outline-primary"
          disabled={false}
          key="Ford"
          onClick={[Function]}
          type="button"
          value="Ford"
        >
          Ford
        </button>
        <button
          className="btn btn-outline-primary"
          disabled={false}
          key="Nike"
          onClick={[Function]}
          type="button"
          value="Nike"
        >
          Nike
        </button>
        <button
          className="btn btn-outline-primary"
          disabled={false}
          key="Unilever"
          onClick={[Function]}
          type="button"
          value="Unilever"
        >
          Unilever
        </button>
        <button
          className="btn btn-outline-primary"
          disabled={false}
          key="Default"
          onClick={[Function]}
          type="button"
          value="Default"
        >
          Default
        </button>
        <button
          className="btn btn-outline-danger"
          onClick={[Function]}
          type="button"
        >
          Reset
        </button>
      </div>
    </div>
  </div>
  <br />
  <br />
</div>
`;
